File: src/main/java/net/ccbluex/liquidbounce/mcef/listeners/OkHttpProgressInterceptor.java
/*
 * This file is part of LiquidBounce (https://github.com/CCBlueX/LiquidBounce)
 *
 * Copyright (c) 2015 - 2025 CCBlueX
 *
 * LiquidBounce is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * LiquidBounce is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LiquidBounce. If not, see <https://www.gnu.org/licenses/>.
 */
package net.ccbluex.liquidbounce.mcef.listeners;

import okhttp3.Interceptor;
import okhttp3.Response;

/**
 * Stub implementation of OkHttpProgressInterceptor for native GUI migration
 * 
 * This replaces the original MCEF progress interceptor functionality with no-op stubs
 * since the native GUI doesn't require MCEF browser progress tracking.
 */
public class OkHttpProgressInterceptor implements Interceptor {
    
    public interface ProgressListener {
        void onProgress(long bytesRead, long contentLength, boolean done);
    }
    
    private final ProgressListener listener;
    
    public OkHttpProgressInterceptor(ProgressListener listener) {
        this.listener = listener;
    }
    
    @Override
    public Response intercept(Chain chain) throws java.io.IOException {
        okhttp3.Request request = chain.request();
        Response response = chain.proceed(request);
        
        // In the original implementation, this would track download progress
        // For the stub, we just pass through the response without progress tracking
        
        return response;
    }
}

File: src/main/kotlin/net/ccbluex/liquidbounce/utils/http/OkHttpProgressInterceptor.kt
/*
 * This file is part of LiquidBounce (https://github.com/CCBlueX/LiquidBounce)
 *
 * Copyright (c) 2015 - 2025 CCBlueX
 *
 * LiquidBounce is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * LiquidBounce is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LiquidBounce. If not, see <https://www.gnu.org/licenses/>.
 */
package net.ccbluex.liquidbounce.utils.http

import okhttp3.Interceptor
import okhttp3.Response
import okio.Buffer
import okio.ForwardingSource
import okio.Source
import okio.buffer
import java.io.IOException

/**
 * OkHttp progress interceptor replacement for the MCEF version.
 * Tracks download progress for HTTP requests.
 */
class OkHttpProgressInterceptor(private val progressListener: ProgressListener) : Interceptor {

    fun interface ProgressListener {
        fun update(bytesRead: Long, contentLength: Long, done: Boolean)
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val originalResponse = chain.proceed(chain.request())
        val originalBody = originalResponse.body
        return if (originalBody != null) {
            originalResponse.newBuilder()
                .body(ProgressResponseBody(originalBody, progressListener))
                .build()
        } else {
            originalResponse
        }
    }

    private class ProgressResponseBody(
        private val responseBody: okhttp3.ResponseBody,
        private val progressListener: ProgressListener
    ) : okhttp3.ResponseBody() {

        private var bufferedSource: okio.BufferedSource? = null

        override fun contentType() = responseBody.contentType()

        override fun contentLength() = responseBody.contentLength()

        override fun source(): okio.BufferedSource {
            if (bufferedSource == null) {
                bufferedSource = source(responseBody.source()).buffer()
            }
            return bufferedSource!!
        }

        private fun source(source: Source): Source {
            return object : ForwardingSource(source) {
                var totalBytesRead = 0L

                @Throws(IOException::class)
                override fun read(sink: Buffer, byteCount: Long): Long {
                    val bytesRead = super.read(sink, byteCount)
                    totalBytesRead += if (bytesRead != -1L) bytesRead else 0
                    progressListener.update(
                        totalBytesRead,
                        responseBody.contentLength(),
                        bytesRead == -1L
                    )
                    return bytesRead
                }
            }
        }
    }
}
